pkg/cli/cmd_helpers.go

```diff
@@ -17,69 +17,275 @@ limitations under the License.
+		executionMethodsPostRunEFunc(pluginKeys, subcommands, cfg, msg)
+}
+
+// executionMethodsPreRunEFunc returns a cobra RunE function that loads the configuration
+// and executes inject config, inject resource and pre-scaffold methods.
+func executionMethodsPreRunEFunc(
+	pluginKeys []string,
+	subcommands *[]plugin.Subcommand,
+	cfg store.Store,
+	options *resourceOptions,
+	fs machinery.Filesystem,
+	msg string,
+) func(*cobra.Command, []string) error {
 	return func(*cobra.Command, []string) error {
-		err := cfg.Load()
-		if os.IsNotExist(err) {
+		if err := cfg.Load(); os.IsNotExist(err) {
 			return fmt.Errorf("%s: unable to find configuration file, project must be initialized", msg)
 		} else if err != nil {
 			return fmt.Errorf("%s: unable to load configuration file: %w", msg, err)
 		}
 
-		subcmd.InjectConfig(cfg.Config())
+		var res *resource.Resource
+		if options != nil {
+			options.Domain = cfg.Config().GetDomain() // TODO: offer a flag instead of hard-coding project-wide domain
+			if err := options.validate(); err != nil {
+				return fmt.Errorf("%s: unable to create resource: %w", msg, err)
+			}
+			res = options.newResource()
+		}
+
+		// Inject config method.
+		subcommandsCopy := make([]plugin.Subcommand, len(*subcommands))
+		copy(subcommandsCopy, *subcommands)
+		for i, subcommand := range subcommandsCopy {
+			if subcmd, requiresConfig := subcommand.(plugin.RequiresConfig); requiresConfig {
+				if err := subcmd.InjectConfig(cfg.Config()); err != nil {
+					var exitError plugin.ExitError
+					if errors.As(err, &exitError) {
+						fmt.Printf("skipping %q: %s\n", pluginKeys[i], exitError.Reason)
+						*subcommands = append((*subcommands)[:i], (*subcommands)[i+1:]...)
+					} else {
+						return fmt.Errorf("%s: unable to inject the configuration to %q: %w", msg, pluginKeys[i], err)
+					}
+				}
+			}
+		}
+
+		// Inject resource method.
+		if res != nil {
+			subcommandsCopy = make([]plugin.Subcommand, len(*subcommands))
+			copy(subcommandsCopy, *subcommands)
+			for i, subcommand := range subcommandsCopy {
+				if subcmd, requiresResource := subcommand.(plugin.RequiresResource); requiresResource {
+					if err := subcmd.InjectResource(res); err != nil {
+						var exitError plugin.ExitError
+						if errors.As(err, &exitError) {
+							fmt.Printf("skipping %q: %s\n", pluginKeys[i], exitError.Reason)
+							*subcommands = append((*subcommands)[:i], (*subcommands)[i+1:]...)
+						} else {
+							return fmt.Errorf("%s: unable to inject the resource to %q: %w", msg, pluginKeys[i], err)
+						}
+					}
+				}
+			}
+			if err := res.Validate(); err != nil {
+				return fmt.Errorf("%s: created invalid resource: %w", msg, err)
+			}
+		}
+
+		// Pre-scaffold method.
+		subcommandsCopy = make([]plugin.Subcommand, len(*subcommands))
+		copy(subcommandsCopy, *subcommands)
+		for i, subcommand := range subcommandsCopy {
+			if subcmd, hasPreScaffold := subcommand.(plugin.HasPreScaffold); hasPreScaffold {
```

Comment:
might want to consider a helper for this pattern, considering it's repeated multiple times across this file -- could abstract over the error handling that skips certain subcommands/copying/etc, and just have something like

```go
err := ProcessSubcommands(subcommands, func(cmd Subcommand) error {
  subcmd, hasPreScaffold := cmd.HasPreScaffold
  if !hasPreScaffold {
    return nil
  }
  return subcmd.PreScaffold(fs)
})
```